File: public/codigos/CNPJ.vb
Imports System.Text.RegularExpressions

Public Class CnpjValidador
    Public Shared Function CalculaCnpjDv(cnpj As String) As String
        cnpj = Regex.Replace(cnpj, "[^A-Za-z0-9]", "").ToUpper()

        If cnpj.Length < 12 Then
            Throw New ArgumentException("CNPJ inválido (tamanho incorreto)")
        End If

        Dim cnpjBase As String = cnpj.Substring(0, 12)
        Dim pesosDV() As Integer = {6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2}

        Dim somatorioDV1 As Integer = 0
        Dim somatorioDV2 As Integer = 0

        For i As Integer = 0 To 11
            Dim asciiDigito As Integer = Asc(cnpjBase(i)) - 48
            somatorioDV1 += asciiDigito * pesosDV(i + 1)
            somatorioDV2 += asciiDigito * pesosDV(i)
        Next

        Dim restoDV1 As Integer = somatorioDV1 Mod 11
        Dim dv1 As Integer = If(restoDV1 < 2, 0, 11 - restoDV1)

        somatorioDV2 += dv1 * pesosDV(12)

        Dim restoDV2 As Integer = somatorioDV2 Mod 11
        Dim dv2 As Integer = If(restoDV2 < 2, 0, 11 - restoDV2)

        Return dv1.ToString() & dv2.ToString()
    End Function

    Public Shared Function ValidaCnpj(cnpj As String) As Boolean
        cnpj = Regex.Replace(cnpj, "[^A-Za-z0-9]", "").ToUpper()

        If cnpj.Length <> 14 OrElse cnpj.Substring(0, 12) = "000000000000" Then
            Return False
        End If

        Dim cnpjBase As String = cnpj.Substring(0, 12)
        Dim dvFornecido As String = cnpj.Substring(12, 2)

        Try
            Dim dvCalculado As String = CalculaCnpjDv(cnpjBase)
            Return dvFornecido = dvCalculado
        Catch
            Return False
        End Try
    End Function
End Class

File: public/codigos/CNPJ.cs
using System;
using System.Text.RegularExpressions;

namespace CnpjUtils
{
    public static class CnpjValidador
    {
        public static string CalculaCnpjDv(string cnpj)
        {
            cnpj = Regex.Replace(cnpj, "[^A-Za-z0-9]", "").ToUpper();

            if (cnpj.Length < 12)
            {
                throw new ArgumentException("CNPJ inválido (tamanho incorreto)");
            }

            string cnpjBase = cnpj.Substring(0, 12);
            int[] pesosDV = { 6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2 };

            int somatorioDV1 = 0;
            int somatorioDV2 = 0;

            for (int i = 0; i < 12; i++)
            {
                int asciiDigito = cnpjBase[i] - 48;
                somatorioDV1 += asciiDigito * pesosDV[i + 1];
                somatorioDV2 += asciiDigito * pesosDV[i];
            }

            int restoDV1 = somatorioDV1 % 11;
            int dv1 = restoDV1 < 2 ? 0 : 11 - restoDV1;

            somatorioDV2 += dv1 * pesosDV[12];

            int restoDV2 = somatorioDV2 % 11;
            int dv2 = restoDV2 < 2 ? 0 : 11 - restoDV2;

            return $"{dv1}{dv2}";
        }

        public static bool ValidaCnpj(string cnpj)
        {
            cnpj = Regex.Replace(cnpj, "[^A-Za-z0-9]", "").ToUpper();

            if (cnpj.Length != 14 || cnpj.Substring(0, 12) == "000000000000")
            {
                return false;
            }

            string cnpjBase = cnpj.Substring(0, 12);
            string dvFornecido = cnpj.Substring(12, 2);

            try
            {
                string dvCalculado = CalculaCnpjDv(cnpjBase);
                return dvFornecido == dvCalculado;
            }
            catch
            {
                return false;
            }
        }
    }
}
